src/DotNetty.Common/Utilities/ICharSequence.cs

```diff
@@ -0,0 +1,28 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace DotNetty.Common.Utilities
+{
+    using System.Collections.Generic;
+
+    public interface ICharSequence : IReadOnlyList<char>
```

Comment:
It seems like most of implementations except of AppendableCharSequence are immutable. AsciiString has mutable "computed" values (string and hash) which are mutated by ArrayChanged which is never used. So they all can be made structs. This change by itself probably won't have much perf impact but if in addition to that all methods accepting ICharSequence will be converted to accept generic parameter T: CharSequence (which I realize will be HUGE change) I wonder how much it would reduce GC pressure.